app/components/inputs/PathBrowseInput.js

```diff
@@ -0,0 +1,64 @@
+const electron = require("electron");
+const dialog = electron.remote.dialog;
+const mainWindow = electron.remote.getCurrentWindow();
+const ipc = electron.ipcRenderer;
+
+import PathInput from "./PathInput";
+import { PathButton } from "../buttons";
+
+let pathListener;
+
+@autobind
+class PathBrowseInput extends React.Component {
+
+  constructor(props) {
+    super(props);
+    this.state = { path: "" };
+  }
+
+  componentDidMount() {
+
+    let self = this;
+    pathListener = function (event, data) {
+      self.setState({ path: data });
+      self.props.onChange(data);
+    };
+
+    ipc.on(this.props.id, pathListener);
+  }
+
+  componentWillUnmount() {
+    ipc.removeListener(this.props.id, pathListener);
+  }
+
+  selectDirectory() {
+    dialog.showOpenDialog(mainWindow, {
+      properties: [this.props.type === "directory" ? "openDirectory" : "openFile"]
+    }, this.directorySelectorCallback);
+  }
+
+  directorySelectorCallback(filenames) {
+    if (filenames && filenames.length > 0) {
+      mainWindow.webContents.send(this.props.id, filenames[0]);
+    }
+  }
+
+  render() {
+    return (
+      <div className={"path-input-and-button"}>
+        <PathInput
+          value={this.props.value}
+          onChange={(e) => {
+            this.props.onChange(e.target.value);
+            this.setState({ path: e.target.value });
+          }}
```

Comment:
This isn't wrong, but just to maintain a general stylistic convention, could you move all this code into an `onChange()` method for this class? 

On the rest of the project, if the event more than a simple, single method call or assignment we usually write a separate handler for it (so the `render()` method is simpler).

Reply:
got it